File: arrays-hashing/two-sum/ts.py
from typing import List

nums = [3, 2, 4]
target = 6

# Try 1 - Brute force
"""
1. we set n to the length of the entire nums array, because we want to use this to iterate over
2. we start the for loop in
3. we start a second for loop with the index incremented by 1
4. we check if the index of nums[i] + nums[j] is equal to the target number. 
5. if the two indicies add up to the target, we return the indicies as an array [i, j]


time complexity:
O(n^2) because there are nested loops iterating through all possible pairs of elements.

Space:
o(1)

"""
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)
        for i in range(n):
            for j in range(i + 1, n):
                if nums[i] + nums[j] == target:
                    return [i, j]
        return []
    
solution = Solution()
result = solution.twoSum(nums, target)
print(result)

File: arrays-hashing/two-sum/ts.js
var nums = [3, 2, 4, 8, 10, 21];
var target = 23;
/**
 * Try 1: brute force
 * 1. define a function that takes in nums array and the target number
 * 2. iterate through the nums array
 * 3. start a nested for loop, increment the starting index by 1 (i + 1)
 * 4. if the sum of the two indices match the target number, return the indices in an array
 *
 *
 */
function twoSum(nums, target) {
    for (var i = 0; i < nums.length; i++) {
        for (var j = i + 1; j < nums.length; j++) {
            if (nums[i] + nums[j] === target) {
                console.log([i, j]);
                return [i, j];
            }
        }
    }
    return [];
}
// twoSum(nums, target);
/**
 * Try 2: hash map
 *
 */
function twoSum2(nums, target) {
    var map = new Map();
    for (var i = 0; i < nums.length; i++) {
        var diff = target - nums[i];
        if (map.has(diff)) {
            return [map.get(diff), i];
        }
        map.set(nums[i], i);
    }
    console.log(map);
    return [];
}
twoSum2(nums, target);
